README
# Bio::ViennaNGS 0.19.2

Bio::ViennaNGS is a distribution of Perl modules and utilities for
building efficient Next-Generation Sequencing (NGS) analysis
pipelines. It covers various aspects of NGS data analysis, including
(but not limited to) conversion of sequence annotation, evaluation of
mapped data, expression quantification and visualization.

Bio::ViennaNGS is shipped with a complementary set of modules, classes and
roles:

- Bio::ViennaNGS::AnnoC
  A Moose interface for storage and conversion of sequence annotation
  data.

- Bio::ViennaNGS::Bam
  Routines for high-level manipulation of BAM files.

- Bio::ViennaNGS::Bed 
  A Moose interface for manipulation of genomic interval data in BED
  format.

- Bio::ViennaNGS::BedGraphEntry
  A Moose interface for storing genomic cvoerage and interval data in
  bedGraph format.

- Bio::ViennaNGS::Expression
  A Moose interface for computation of normalized gene / transcript
  expression based on read counts.

- Bio::ViennaNGS::ExtFeature
  A Moose wrapper for extended BED6 elements.

- Bio::ViennaNGS::Fasta
  Routines for accessing genomic sequences implemented through a Moose
  interface to Bio::DB::Fasta.

- Bio::ViennaNGS::Feature
  A Moose-based BED6 wrapper.  

- Bio::ViennaNGS::FeatureChain
  Yet another Moose class for chaining gene annotation features. 

- Bio::ViennaNGS::FeatureInterval
  A Moose interface for handling elementary genomic intervals,
  corresponding to BED3.

- Bio::ViennaNGS::FeatureIO
  A Moose interface for easy input/output operations on common feature
  annotation file formats.

- Bio::ViennaNGS::FeatureLine
  An abstract Moose class for combining several
  Bio::ViennaNGS::FeatureChain objects.

- Bio::ViennaNGS::MinimalFeature
  A Moose interface for handling elementary gene annotation, corresponding
  to BED4.

- Bio::ViennaNGS::Peak 
  A Moose interface for identification and characterization of (coverage /
  expression) peaks in RNA-seq data.

- Bio::ViennaNGS::SpliceJunc
  A collection of routines for alternative splicing analysis.

- Bio::ViennaNGS::Subtypes
  Moose subtypes for internal usage.

- Bio::ViennaNGS::Tutorial
  A comprehensive tutorial of the Bio::ViennaNGS core routines with
  real-world NGS data.

- Bio::ViennaNGS::UCSC
  Routines for visualization of genomics data with the UCSC genome
  browser.

- Bio::ViennaNGS::Util
  A collection of wrapper routines for commonly used third-party NGS
  utilities as well as a set of utility functions.


## UTILITIES

In addition, Bio::ViennaNGS comes with a collection of utility
programs for accomplishing routine tasks often required in NGS data
processing. These utilities serve as reference implementation of the
routines implemented in the (sub)modules:

assembly_hub_constructor.pl:
The UCSC genome browser offers the possibility to visualize any
organism (including organisms that are not included in the standard
UCSC browser bundle) through hso called 'Assembly Hubs'. This script
constructs Assembly Hubs from genomic sequence and annotation data.

bam_split.pl: 
Split (paired-end and single-end) BAM alignment files by strand and compute
statistics. Optionally create BED output, as well as normalized bedGraph
and bigWig files for coverage visualization in genome browsers (see
dependencies on third-patry tools below).

bam_to_bigWig.pl:
Produce bigWig coverage profiles from (aligned) BAM files, explicitly
considering strandedness. The most natural use case of this tool is to
create strand-aware coverage profiles in bigWig format for genome browser
visualization.

bam_uniq.pl:
Extract unique and multi mapping reads from BAM alignment files and create
a separate BAM file for both uniqe (.uniq.) and multi (.mult.) mappers.

bed2bedGraph.pl:
Convert BED files to (strand specific) bedGraph files, allowing
additional annotation and automatic generation of bedGraph files which can
easily be converted to big-type files for easy UCSC visualization.

bed2nt2aa.pl:
Extract sequence intervals from Fasta files, provided as BED6
records. Optionally translate nucleotide into amino acid sequences.

bed62bed12.pl:
A simple BED6 -> BED12 converter

fasta_multigrep.pl:
Extract individual (whole) sequences from a multi Fasta file

fasta_subgrep.pl:
Extract subsequences from (multi) Fasta files.

fasta_regex.pl:
Extract sequence motifs from (multi) Fasta files provided as regular
expression.

gff2bed.pl:
Convert RefSeq GFF3 annotation files to BED12 format. Individual BED12
files are created for each feature type (CDS/tRNA/rRNA/etc.). Tested with
RefSeq bacterial GFF3 annotation.  

kmer_analysis.pl:
Count k-mers of predefined length in FastQ and Fasta files

newUCSCdb.pl:
Create a new genome database to a locally installed instance of the UCSC
genome browser in order to add a novel organism for visualization.

normalize_multicov.pl: 
Compute normalized expression data in TPM and RPKM from (raw) read counts
in bedtools multicov format. TPM reference: Wagner et al, Theory
Biosci. 131(4), pp 281-85 (2012).

rnaseq_peakfinder.pl:
Find and characterize peaks/enriched regions of certain size and coverage
in RNA-seq data.

sj_visualizer.pl:
Convert splice junctions from mapped RNA-seq data in segemehl BED6 splice
junction format to BED12 for easy visualization in genome browsers.

splice_site_summary.pl:
Identify and characterize splice junctions from RNA-seq data by
intersecting them with annotated splice junctions.

track_hub_constructor.pl:
Analogous to assembly_hub_constructor.pl, construct a Track Hub for an
organism listed in the UCSC Genome Browser.

trim_fastq.pl:
Trim sequence and quality string fields in a Fastq file by user
defined length.


## TUTORIAL

See Bio::ViennaNGS::Tutorial 


## INSTALLATION

We have compiled a comprehensive `ViennaNGS Installation HOWTO` which
covers the setup of third party and Perl dependencies in detail:

http://rna-seq.at/blog/2015/03/03/viennangs-installation-howto/

If you have already installed all dependencies, the ViennaNGS distribution
can be installed with the following commands:

   perl Makefile.PL
   make
   make test
   make install

However, we highly recommend using *cpanminus* for installing ViennaNGS. If
you have cpanm available on your system, simply type

  cpanm Bio::ViennaNGS

This command will grab the lastest ViennaNGS release from CPAN together
with all Perl dependencies and install the suite automatically (provided
you have samtools <= 0.1.19 installed, see the HOWTO link above).


## SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Bio::ViennaNGS

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-ViennaNGS

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Bio-ViennaNGS

    CPAN Ratings
        http://cpanratings.perl.org/d/Bio-ViennaNGS

    Search CPAN
        http://search.cpan.org/dist/Bio-ViennaNGS


## DEPENDENCIES

Bio::ViennaNGS depends on a set of third-party tools and libraries which
are required for specific filtering and file format conversion tasks as
well as for building internally used Perl modules:

* bedtools2 >= 2.17 (https://github.com/arq5x/bedtools2) 
* bedGraphToBigWig, fetchChromSizes, faToTwoBit from the UCSC Genome
  Browser applications (http://hgdownload.cse.ucsc.edu/admin/exe/)
* the R Statistics software (http://www.r-project.org/)
* samtools <= v0.1.19 for building Bio::DB::Sam. Please note that more
  recent HTSlib-based versions of samtools will not work with Bio::DB::Sam 

Please ensure that all third-party utilities are available on your system
and accessible to the Perl interpreter.

Moreover, Bio::ViennaNGS depends on a ste of Perl packages that are not
shipped with the Perl core distribution, most notably Bio::Perl >=
1.00690001 and Bio::DB::Sam >= 1.37.


## SOURCE AVAILABILITY

This source is available on Github: 
              https://github.com/mtw/Bio-ViennaNGS


## VIENNANGS PAPER

If the Bio::ViennaNGS suite is useful for your work or if you use any
component of Bio::ViennaNGS in a custom pipeline, please cite the following
publication:

"ViennaNGS - A toolbox for building efficient next-generation sequencing
analysis pipelines"
Michael T. Wolfinger, Joerg Fallmann, Florian Eggenhofer and Fabian Amman
F1000Research 4:50 (2015)
DOI: http://dx.doi.org/10.12688/f1000research.6157.2


## NOTES

The Bio::ViennaNGS suite is actively developed and tested on different
flavours of Linux and Mac OS X. We have taken care of writing
platform-independent code that should run out of the box on most UNIX-based
systems, however we do not have access to machines running Microsoft
Windows. As such Microsoft Windows compatibility is not tested.


## BUGS

Please report bugs through the Github issue tracker at
https://github.com/mtw/Bio-ViennaNGS/issues


## AUTHORS

Michael T. Wolfinger <michael@wolfinger.eu>
Joerg Fallmann <fall@tbi.univie.ac.at>
Florian Eggenhofer <florian.eggenhofer@tbi.univie.ac.at>
Fabian Amman <fabian@tbi.univie.ac.at>


## COPYRIGHT AND LICENCE

Copyright (C) 2014-2017 Michael T. Wolfinger <michael@wolfinger.eu>

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself, either Perl version 5.10.0 or, at your
option, any later version of Perl 5 you may have available.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.


Changes
Revision history for Perl extension Bio::ViennaNGS.

0.01  Fri Sep 13 12:02:47 2013
	- original version; created by h2xs 1.23 with options
		-XA -n ViennaNGS
	- added get_stranded_subsequence()

0.02  Mon Nov 4 2013
	- added split_bam()
        - added scripts/bam_split.pl

0.03  Tue Nov 5 2013
    	- added bam2bw()
	- updated README

0.04  Tue Nov 26 2013
	- added scripts/DEXSeq_gff2bed.pl
	- updated README

0.05  Thu Apr 03 2014
	- added bed2bw()
	- updated and extended bam_split()
	- updated scripts/bam_split.pl
	- adjusted perl shebang in /scripts

0.06  Fri Sep 19 2014
	- adjusted perl shebang in /scripts
	- added normalize_multicov.pl in /scripts. For that purpose:
	- added computeTPM()
	- added parse_multicov()
	- added write_multicov()
	- added featCount_data() accessor for @featCount AoHoH 
	- added normalization support for bigWigs generated from scripts/bam_split.pl
	- refactored split_bam()
	- intermediate bedGraph files are now removed in bed2bw()

0.07  Mon Sep 29 2014
	- minimum version of Bio::Perl is 1.006924
	- minimum version of Bio::DB::Sam is 1.39
	- added scripts/splice_site_summary.pl
	- added scripts/assembly_hub_constructer.pl prototype
	- added bed2bigBed()
	- use IPC::Cmd for running external tools in bed2bigBed()
	- use Carp for error handling
	- refactored bam2bw()
	- refactored bed2bw()
	- removed scripts/ensembl_fa2bed.pl

0.08  Tue Sep 30 2014
	- added tests for third-party utilities
	- use Carp for error handling throughout the module
	- updated documentation

0.09  Thu Oct 2 2014
	- changed namespace to Bio::ViennaNGS
	- adjusted utilities to use new namespace
	- everything is exported via @EXPORT_OK now
	- updated README

0.10  Thu Nov 27 2014
	- requires Perl >= 5.12.0
	- added sortbed()
	- added uniquify_bam()
	- added scripts/bam_uniq.pl
	- added scripts/sj_visualizer.pl
	- refactored scripts/motiffinda.pl
	- removed development version of scripts/assembly_hub_constructer.pl
	- made scripts/gff2bed.pl work with OO Bio::ViennaNGS::AnnoC 
	- removed get_stranded_subsequence (now lives in Bio::ViennaNGS::Fasta)
	- updated POD

0.11 Wed Dec 10 2014
	- integrated Bio::ViennaNGS::AnnoC
	- integrated Bio::ViennaNGS::Fasta 
	- removed obsolete bam2bw()
	- added scripts/bam_to_bigwig.pl
	- added scripts/bed2bedGraph.pl
	- added scripts/extend_bed.pl
	- added scripts/kmer_analysis.pl
	- added scripts/MEME_XML_motif_extractor.pl
	- added scripts/trim_fastq.pl
	- refactored scripts/bam_to_bigwig.pl
	- refactored scripts/splice_site_summary.pl
	- refactored scripts/bam_split.pl
	- refactored scripts/gff2bed.pl
	- added Moose-based classes for feature handling	
	- renamed bed2bw() to bam_or_bed2bw() and refactored it accordingly to
	  properly handle BAM files
	- added '-split' option to 'genomeCoverageBed' call in
	  bam_or_bed2bw()
	- updated MakeMaker scripts (MANIFEST and Makefile.PL)
	- minimum version of Bio::Root::Version is now 1.00690001
	- added some tests
	- ship demo files in share/ folder
	- updated POD and README
	
0.12 Wed Feb 11 2015
	- requires Perl >= 5.10.0
	- minimum version of Bio::DB::Sam is 1.37
	- integrated Bio::ViennaNGS::Bam
	- integrated Bio::ViennaNGS::BamStat
	- integrated Bio::ViennaNGS::BamStatSummary
	- integrated Bio::ViennaNGS::Bed
	- integrated Bio::ViennaNGS::ExtFeature
	- integrated Bio::ViennaNGS::SpliceJunc
	- integrated Bio::ViennaNGS::Tutorial
	- integrated Bio::ViennaNGS::UCSC
 	- integrated Bio::ViennaNGS::FeatureLine
	- added scripts/assembly_hub_generator.pl
	- added scripts/bam_quality_stat.pl
	- added scripts/Tutorial_pipeline00.pl
	- added scripts/Tutorial_pipeline01.pl
	- added scripts/Tutorial_pipeline02.pl
	- added scripts/Tutorial_pipeline03.pl
	- added scripts/track_hub_constructor.pl
	- renamed Bio::ViennaNGS.pm to Bio::ViennaNGS::Util.pm
	- migrated BAM specific code from Bio::ViennaNGS::Util 
	  to Bio::ViennaNGS::Bam
	- migrated expression/normalization code from Bio::ViennaNGS::Util
	  to Bio::ViennaNGS::Expression
	- refactored scripts/normalize_multicov.pl
	- removed scripts/motiffinda.pl from distribution
	- added static templates for the Assembly and Track Hub generator
	- updated POD and README 

0.13 Thu Feb 14 2015
	- fixed an issue in Bio::ViennaNGS::FeatureChain
	- updated POD

0.14 Tue Feb 24 2015
	- updated mkdircheck() 
	- updated rmdircheck()
	- fixed directory creation in bam_split.pl
	- updated POD in Tutorials

0.15 Mon Jul 6 2015
	- fixed path creation in bed_or_bam2bw()
	- fixed scalar initialization bug
        - fixed bug in Bio::ViennaNGS::BamStat
	- scripts/bam_to_bigwig.pl can now produce normalized bigWigs
	- added functionality to compute RPKM in Bio::ViennaNGS::Expression
	- added tests for UCSC.pm
	- added tests for SpliceJunc.pm

0.16 Tue Oct 27 2015
	- added Bio::ViennaNGS::FeatureInterval
	- added Bio::ViennaNGS::BedGraphLine
	- added Bio::ViennaNGS::Peak
	- make Bio::ViennaNGS::MinimalFeature subclass of
	  Bio::ViennaNGS::FeatureInterval
	- added scripts/rnaseq_peakfinder.pl
	- added scripts/bed2nt2aa.pl
	- updated POD and README

0.17 Fri Jun 09 2017
	- added Bio::ViennaNGS::FeatureIO
	- added Bio::ViennaNGS::Subtypes
	- refactored Bio::ViennaNGS::Fasta to transparently coerce input
	  data into a Bio::DB::Fasta object
	- refactored executables in scripts/ folder to new
	  Bio::ViennaNGS::Fasta interface
	- changed output format in scripts/bed2nt2aa.pl to csv
	- removed parse_bed6() and extend_chain() from Bio::ViennaNGS::Util
	  (bed I/O is now available in Bio::ViennaNGS::FeatureIO and chain
	   extension is available in bedtools)
	- added uniquify_bam2() in Bio::ViennaNGS::Bam
	- implemented correct handling of unmapped reads in
	  Bio::ViennaNGS::Bam
	- replace '~' with $HOME in mkdircheck() within
          Bio::ViennaNGS::Util
	- fixed output path creation in rnaseq_peakfinder.pl,
	  splice_site_summary.pl
	- refactored scripts/bam_uniq.pl
	- refactored scripts/bed2nt2aa.pl
	- added scripts/fasta_subgrep.pl
	- added scripts/fasta_multigrep.pl

Makefile.PL
use 5.010000;
use lib 'inc';
use ExtUtils::MakeMaker;
use strict;
use warnings;
use File::ShareDir::Install;

install_share dist => 'share';

my @scripts = qw (assembly_hub_constructor.pl bam_split.pl bam_to_bigwig.pl bam_uniq.pl bed2bedGraph.pl bed2nt2aa.pl bed62bed12.pl gff2bed.pl fasta_subgrep.pl fasta_multigrep.pl fasta_regex.pl kmer_analysis.pl newUCSCdb.pl normalize_multicov.pl rnaseq_peakfinder.pl sj_visualizer.pl splice_site_summary.pl track_hub_constructor.pl trim_fastq.pl Tutorial_pipeline03.pl);
WriteMakefile(
	      'NAME'            => 'Bio::ViennaNGS',
	      'AUTHOR'          => q{Michael T. Wolfinger},
	      'VERSION_FROM'    => 'lib/Bio/ViennaNGS.pm',
	      'ABSTRACT_FROM'   => 'lib/Bio/ViennaNGS.pm',
	      'LICENSE'         => 'perl',
	      'PREREQ_PM'       => {
				    "Getopt::Long"	           => '2.0',
				    "constant"                     => '0',
				    "Carp" 		           => '0',
				    "Cwd"                          => '0',
				    "Bio::Perl"                    => '0',
				    "Bio::Root::Version"           => '1.00690001',
				    "Bio::SeqUtils"                => '0',
				    "Bio::DB::Sam"                 => '1.37',
				    "Bio::DB::Fasta"               => '0',
				    "Bio::Tools::GFF"              => '0',
				    "Data::Dumper"                 => '0',
				    "Env"                          => '0',
				    "Exporter"                     => '0',
				    "File::Basename"     	   => '0',
				    "File::Find"                   => '0',
				    "File::Path"                   => '0',
				    "File::Share"                  => '0',
				    "File::Temp" 	           => '0',
				    "File::Slurp"                  => '0',
				    "FindBin"                      => '0',
				    "IPC::Cmd" 		           => '0',
				    "lib"                          => '0',
				    "List::Util"                   => '0',
				    "Moose"      		   => '0',
				    "MooseX::Clone"                => '0',
				    "Moose::Util::TypeConstraints" => '0',
				    "Moose::Role"                  => '0',
				    "namespace::autoclean"         => '0',
				    "Path::Class" 	           => '0',
				    "Math::Round"                  => '0',
				    "Params::Coerce"               => '0',
				    "Pod::Usage" 	           => '0',
				    "PerlIO::gzip"                 => '0',
				    "POSIX" 		           => '0',
				    "strict"                       => '0',
				    "Template"                     => '0',
				    "Test::Files"                  => '0',
				    "Test::File::Contents"         => '0',
				    "Test::Deep"                   => '0',
				    "Tie::Hash::Indexed"           => '0',
				    "version"                      => '0',
				    "warnings"                     => '0',
				   },
	      'MIN_PERL_VERSION'          => '5.010',
	      'EXE_FILES'                 => [ map { "scripts/$_" } @scripts ],
	      'CONFIGURE_REQUIRES'        => {
					      'ExtUtils::MakeMaker' => 0,
					     },
	      'BUILD_REQUIRES' => {
				   "Test::More" => 0,
				   'Test::Prereq' => 0,

				  },
	      dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
	      clean => { FILES => 'Bio-ViennaNGS-*' },
	     );

package MY;
use File::ShareDir::Install qw(postamble);
